packages/venia-ui/lib/drivers/adapter.js

```diff
@@ -46,44 +47,37 @@ const storage = new BrowserPersistence();
  */
 const VeniaAdapter = props => {
     const { apiBase, apollo = {}, children, store } = props;
+    const [initialized, setInitialized] = useState(false);
 
-    const cache = apollo.cache || preInstantiatedCache;
-    const link = apollo.link || VeniaAdapter.apolloLink(apiBase);
-
-    const persistor = new CachePersistor({
-        cache,
-        storage: window.localStorage,
-        debug: process.env.NODE_ENV === 'development'
-    });
-
-    let apolloClient;
-    if (apollo.client) {
-        apolloClient = apollo.client;
-    } else {
-        apolloClient = new ApolloClient({
+    const apolloClient = useMemo(() => {
+        const cache = apollo.cache || preInstantiatedCache;
+        const link = apollo.link || VeniaAdapter.apolloLink(apiBase);
+        const client = apollo.client || new ApolloClient({ cache, link });
+        const persistor = new CachePersistor({
             cache,
-            link
+            storage: window.localStorage,
+            debug: process.env.NODE_ENV === 'development'
         });
-        apolloClient.apiBase = apiBase;
-    }
-
-    apolloClient.persistor = persistor;
 
-    const [initialized, setInitialized] = useState(false);
+        return Object.assign(client, { apiBase, persistor });
+    }, [apiBase, apollo]);
 
+    // perform blocking async work here
     useEffect(() => {
-        async function initialize() {
-            // On load, restore the persisted data to the apollo cache and then
-            // allow rendering. You can do other async blocking stuff here.
-            if (persistor) {
-                await persistor.restore();
-            }
+        if (initialized) return;
+
+        // immediately invoke this async function
+        (async () => {
+            // restore persisted data to the Apollo cache
+            await apolloClient.persistor.restore();
+
+            // attach the Apollo client to the Redux store
+            attachClient(apolloClient);
```

Comment:
This is the only real change to this file. Ideally, we would create the Apollo client at the same time we create the Redux store and provide it directly to the middleware, but since we've delegated client creation to `Adapter`, we have to wait and add it here.